intel/custom_resource.py

```diff
@@ -0,0 +1,213 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from . import util
+from kubernetes.client.rest import ApiException as K8sApiException
+import datetime
+import json
+import logging
+import time
+
+
+# Example usage:
+#
+# from kubernetes import config as k8sconfig, client as k8sclient
+#
+# k8sconfig.load_kube_config()
+#
+# v1beta = k8sclient.ExtensionsV1beta1Api()
+#
+# apple_type = CustomResourceDefinitionType(
+#               v1beta, "intel.com", "apple", ["ap"])
+#
+# apple1 = apple_type.create("apple1")
+# apple1.body["spec"]["count"] = 5
+# apple1.save()
+#
+# apple2 = apple_type.create("apple2")
+# apple2.body["spec"]["count"] = 10
+# apple2.save()
+
+
+class CustomResourceDefinitionType():
+    def __init__(self, api, group, name, short_names,
+                 version="v1", scope="Namespaced"):
+        assert (api is not None)
+        assert (group is not None)
+        assert (name is not None)
+        assert (short_names is not None)
+        assert (version is not None)
+        assert (scope is not None)
+
+        self.api = api
+        self.group = group
+        self.kind_name = str(name).capitalize()
+        self.singular_name = name
+        self.plural_name = "{}s".format(name)
+        self.short_names = short_names
+        self.type_version = version
+        self.type_scope = scope
+
+    def save(self):
+
+        spec = {
+            "group": self.group,
+            "version": self.type_version,
+            "scope": self.type_scope,
+            "names": {
+                "kind": self.kind_name,
+                "singular": self.singular_name,
+                "plural": self.plural_name,
+                "shortNames": self.short_names
+            }
+        }
+        body = {
+            "metadata": {"name": ".".join([self.plural_name, self.group])},
+            "spec": spec
+        }
+
+        try:
+            header_params = {
+                'Content-Type': "application/json",
+                'Accept': "application/json"
+            }
+
+            auth_settings = ['BearerToken']
+
+            resource_path = \
+                '/apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions'
+            self.api.api_client.call_api(
+                resource_path,
+                'POST',
+                header_params,
+                body=body,
+                auth_settings=auth_settings)
+
+        except K8sApiException as e:
+            if json.loads(e.body)["reason"] != "AlreadyExists":
+                raise e
+
+        # Wait until resource type is ready.
+        while not self.exists():
+            logging.info(
+                "waiting 1 second until third party resource is ready")
+            time.sleep(1)
+
+    def exists(self, namespace="default"):
+        header_params = {
```

Comment:
If using the same variable value in multiple places, extract it and reuse it later